code/modules/modular_computers/file_system/computer_file.dm

```diff
@@ -34,8 +43,118 @@ var/global/file_uid = 0
 	if(metadata)
 		temp.metadata = metadata.Copy()
 	if(rename)
-		temp.filename = filename + "(Copy)"
+		temp.filename = filename + copy_string
 	else
 		temp.filename = filename
 	temp.filetype = filetype
-	return temp
\ No newline at end of file
+	temp.read_access = read_access
+	temp.write_access = write_access
+	temp.mod_access = mod_access
+	return temp
+
+/datum/computer_file/proc/get_file_perms(var/list/accesses, var/mob/user)
+	. = 0
+	if(!accesses || (isghost(user) && check_rights(R_ADMIN, 0, user))) // No access list past means internal usage, so grant full access. Also override for use by admin ghosts.
+		return (OS_READ_ACCESS | OS_WRITE_ACCESS | OS_MOD_ACCESS)
+	if(!LAZYLEN(read_access) || has_access(read_access, accesses))
+		. |= OS_READ_ACCESS
+	if(!LAZYLEN(write_access) || has_access(write_access, accesses))
+		. |= OS_WRITE_ACCESS
+	if(!LAZYLEN(mod_access) || has_access(mod_access, accesses))
+		. |= OS_MOD_ACCESS
+	
+/datum/computer_file/proc/get_perms_readable()
+	var/list/msg = list()
+	msg += "Permissions for file [filename]:"
+	var/read_perms
+	if(LAZYLEN(read_access))
+		read_perms = english_list(read_access[1])
+	else
+		read_perms = "No access required."
+	msg += "Read access: [read_perms]"
+	var/write_perms
+	if(LAZYLEN(write_access))
+		write_perms = english_list(write_access[1])
+	else
+		write_perms = "No access required."
+	msg += "Write access: [write_perms]"
+	var/mod_perms
+	if(LAZYLEN(mod_access))
+		mod_perms = english_list(mod_access[1])
+	else
+		mod_perms = "No access required."
+	msg += "Permission modification access: [mod_perms]"
+	return msg
+
+/datum/computer_file/proc/change_perms(var/change, var/perm, var/access_key, var/changer_accesses)
+
+	if(!has_access(mod_access, changer_accesses))
+		return FALSE
+
+	switch(perm)
+		if(OS_READ_ACCESS)
+			if(change == "+")
+				if(!LAZYLEN(read_access))
+					read_access = list()
+					read_access += list(list())
+				read_access[1] += access_key
+				return TRUE
+			else if(change == "-")
+				if(!LAZYLEN(read_access)) // There weren't any access requirements to begin with.
+					return TRUE
+				read_access[1] -= access_key
+				if(!length(read_access[1]))
+					read_access[1] = null
+					read_access = null
+				return TRUE
+			else
+				return FALSE // Something unexpected was passed into the change argument.
+
+		if(OS_WRITE_ACCESS)
+			if(change == "+")
+				if(!LAZYLEN(write_access))
+					write_access = list()
+					write_access += list(list())
+				write_access[1] += access_key
+				return TRUE
+			else if(change == "-")
+				if(!LAZYLEN(write_access))
+					return TRUE
+				write_access[1] -= access_key
+				if(!length(write_access[1]))
+					write_access[1] = null
+					write_access = null
+				return TRUE
+			else
+				return FALSE
```

Comment:
Think about a way of writing this that doesn't duplicate code.

Reply:
Done, as far as I was able. Mod access still needs to have separate logic.